cpp/src/arrow/array/array_nested.cc

```diff
@@ -216,12 +216,26 @@ static std::shared_ptr<Array> SliceArrayWithOffsets(const Array& array, int64_t
   return array.Slice(begin, end - begin);
 }
 
+namespace {
+struct FlattenWithRecursion {
+  // Flatten all list-like types array recursively
+  static Result<std::shared_ptr<Array>> Flatten(const Array& array, bool with_recursion,
+                                                MemoryPool* memory_pool);
+};
+}  // namespace
+
 template <typename ListArrayT>
 Result<std::shared_ptr<Array>> FlattenListArray(const ListArrayT& list_array,
+                                                bool with_recursion,
                                                 MemoryPool* memory_pool) {
   const int64_t list_array_length = list_array.length();
   std::shared_ptr<arrow::Array> value_array = list_array.values();
 
+  // If it's a nested-list related array, flatten recursively.
+  if (is_list_like(value_array->type_id()) && with_recursion) {
+    return FlattenWithRecursion::Flatten(*value_array, with_recursion, memory_pool);
+  }
+
```

Comment:
Only the recursive version should delegate to this one, not the other way around. So this function shouldn't gain a boolean parameter.